### ISSUE TITLE
Restore original images from X_test processed using `imagenet_utils.preprocess_input`

### ISSUE THREAD

**Issue by khteh** [keras-team/keras#21829]
I have train, validation and test datasets from the original corpora of images. Some of these images are grascale, some RGB and the grayscale images are converted to RGB using `tf.image.grayscale_to_rgb(image)`. These images are then processed using `tf.keras.applications.resnet_v2.preprocess_input`. I need to evaluate using the test dataset and show the result by plotting the images. How to restore those images which have been processed? The challenge here is to undo/reverse the operation done by both of these library functions.

**Comment by Abhinavexists** [comment]
hi @khteh 

you can refer to https://colab.research.google.com/drive/18lSwduKf-vnEzAeNepF99_jN55vCUwPQ?usp=sharing 
here i have implemented a way to reverse both operations done by library functions

Reversed the ResNetV2 preprocessing by rescaling pixel values from [-1, 1] back to [0, 255] using (preprocessed + 1.0) * 127.5 as this is the reverse linear function of the ResnetV2:  preprocessed = (original/127.5) - 1

**Comment by khteh** [comment]
Doesn't work. Here are the values:
```
input (64, 64, 3): mean: -0.9975636601448059 [-0.9997231960296631, -0.9927104711532593]
gray (64, 64, 1): mean: 0.0 [0, 0]
restored (64, 64, 3): mean: 0.0 [0, 0]
```
`input`: The orignal grayscale image first converted to RGB with `tf.image.grayscale_to_rgb(image)`, then put through `preprocess_input()`

**Comment by khteh** [comment]
I can't run the colab notebook and it is using TF 2.19. If you use `rgb_image_new = rgb_image` would it work?

**Comment by Abhinavexists** [comment]
Hi @khteh 
I have changed the collab access from viewer to editor 
try again

**Comment by Abhinavexists** [comment]
> I can't run the colab notebook and it is using TF 2.19. If you use `rgb_image_new = rgb_image` would it work?

I don't think so as it will only copy the existing tensor value of the image

**Comment by khteh** [comment]
Why do you want to multiply the `preprocess_input` image with `255`?

**Comment by khteh** [comment]
> Hi [@khteh](https://github.com/khteh) I have changed the collab access from viewer to editor try again

<img width="2499" height="651" alt="Image" src="https://github.com/user-attachments/assets/7c095096-adf2-459e-9576-7d9f52307ed0" />

**Comment by khteh** [comment]
`* 255` after `tf.image.grayscale_to_rgb(image)` works. Now the values are:
```
input (64, 64, 3): mean: -0.21290023624897003 [-0.7960784435272217, 0.9215686321258545]
gray (64, 64, 1): mean: 99.4765625 [25, 245]
restored (64, 64, 3): mean: 99.4765625 [25, 245]
```

**Comment by khteh** [comment]
Thanks @Abhinavexists

**Comment by Abhinavexists** [comment]
> Why do you want to multiply the `preprocess_input` image with `255`?

@khteh you want to do it so to fit it again in the RGB scale , once it is done you can then directly apply the linear function

**Comment by khteh** [comment]
> > Why do you want to multiply the `preprocess_input` image with `255`?
> 
> [@khteh](https://github.com/khteh) you want to do it so to fit it again in the RGB scale , once it is done you can then directly apply the linear function

This only applies to the originally grayscale images. My dataset has a mix of both and I can't tell which is which as they are all shuffled and batched to create train, validation and test datasets.

**Comment by khteh** [comment]
Oh, wait. That `* 255` is applied at the processing path. Not de-normalization path. Thanks.
